### PR TITLE
feat(streaming): deanonymize DeepSeek reasoning_content

### PR DESCRIPTION
**Pull request by iamclaude697**
## What
Extend OpenAI streaming parser to deanonymize DeepSeek's `reasoning_content` field. DeepSeek's `deepseek-reasoner` model emits chain-of-thought text on `choices[].delta.reasoning_content` ahead of the answer on `choices[].delta.content`; both can carry PII tokens and both must now be rehydrated.

## Changes
- Extended `openAIDelta` struct with `ReasoningContent` field
- Added separate `reasoningAccum` accumulator in `openAIDeanonymizer`
- `ProcessDataPayload` handles `reasoning_content` alongside `content` (mutually exclusive per chunk; either-or guard, never both)
- `Flush()` drains both accumulators — reasoning first, then content, matching DeepSeek's arrival order
- Synthetic flush chunks re-serialize with the originating field (`reasoning_content` or `content`)

No impact on other OpenAI-compatible providers — the field is only present in DeepSeek responses and serialised as empty/omitted elsewhere.

## Baseline Issues
- `TestDispatchOllamaAsyncWithMockServer` is intermittently flaky on `main` (race on the mock-server port; passes when re-run in isolation). Unrelated to this change.

## Quality Gates
- [x] `make check` passed (lint + test + gosec + govulncheck — all green)
- [x] `go test -race ./...` passed across all packages
- [x] No regressions vs baseline — every existing OpenAI streaming test still passes
- [x] All existing OpenAI streaming tests pass unchanged (`TestOpenAIStreamingTokenSplit`, `TestOpenAIStreamingEOFFlush`, `TestOpenAIStreamingDoneSentinel`, `TestOpenAIStreamingRoleOnlyChunk`, `TestOpenAIStreamingFinishReason`, `TestOpenAIStreamingEmptyChoicesFlush`, `TestOpenAIStreamingVerboseLogging`, `TestOpenAIStreamingViaAggregatorDomain`)
- [x] Coverage delta ≥ 90% on changed files (`streaming_openai.go`: `newOpenAIDeanonymizer` 100%, `ProcessDataPayload` 98%, `Flush` 100%)

## Test Plan
- `TestDeepSeekReasoningContentDeanonymized` — token in `reasoning_content` replaced; content untouched
- `TestDeepSeekReasoningTokenSplit` — token split across two reasoning chunks reassembled and replaced
- `TestDeepSeekReasoningThenContent` — full session with one PII token in reasoning phase and one in content phase; both replaced
- `TestDeepSeekReasoningEOFFlush` — reasoning held in accumulator (no finish chunk) is flushed on EOF
- `TestDeepSeekKeepAliveComment` — `: keep-alive` SSE comment lines interleaved between reasoning chunks do not disrupt accumulation
- `TestOpenAIUnchangedByReasoningField` — explicit regression guard: canonical OpenAI token split still passes for `api.openai.com`

## Test Method Compliance
- Read `docs/test-plans/ai-proxy-test-method.md` in full
- §5 Known Issues: none affect this change (all relate to pack regex/validator interference, not streaming)
- §6 Test Inventory: streaming files are out of the pack-oriented inventory; verified the relevant `streaming_openai_test.go` + race suite (`go test -race ./...`) green

---
_Generated by [Claude Code](https://claude.ai/code/session_011yRdJ8gqY11wLWDLTtNQvb)_

### CODE REVIEW

File: internal/anonymizer/streaming_openai.go [hunk 1]
```diff
@@ -68,15 +75,41 @@ func (o *openAIDeanonymizer) ProcessDataPayload(payload []byte) bool {
...
+	// Neither content nor reasoning_content — role-only or other delta — pass through.
+	if choice.Delta.Content == "" && choice.Delta.ReasoningContent == "" {
 		writePipe(o.opts.pw,
 			[]byte(sseDataPrefix),
 			[]byte(o.opts.replacer.Replace(string(payload))),
 			[]byte("\n"))
 		return true
 	}
 
+	if choice.Delta.ReasoningContent != "" {
```
Comment (laplaque) [inline internal/anonymizer/streaming_openai.go:87]: **MEDIUM — streams are not structurally isolated.**

The reasoning branch returns at line 110, so a chunk carrying both `reasoning_content` and `content` silently drops content. Just removing the `return true` does not fix it: the code mutates `choice.Delta.ReasoningContent` and marshals the input envelope, which still carries `content: Y` unreplaced. Running the content branch afterward would then leak both directions — the reasoning output chunk would carry the raw, unreplaced content field, and the content output chunk would carry the already-replaced reasoning field.

Today's single-field cases are correct only because the unused field is empty in input and `omitempty` hides it. That's accidental correctness, not structural.

**Fix.** Emit each stream through a freshly-constructed envelope so the delta carries only that stream's field:

```go
outEnv := openAIEnvelope{
    ID:     envelope.ID,
    Object: envelope.Object,
    Choices: []openAIChoice{{
        Index:        choice.Index,
        Delta:        openAIDelta{ReasoningContent: replaced},
        FinishReason: choice.FinishReason,
    }},
}
```

Mirror for content. Run both branches (no early return) so a both-fields chunk produces two output chunks. Add a test that asserts (a) a single input chunk with both fields produces two separate output chunks, (b) the reasoning chunk contains only the replaced reasoning text and no content text, (c) the content chunk contains only the replaced content text and no reasoning text.

This also future-proofs against new delta fields (tool_calls, function_calls, etc.) — separation becomes structural rather than dependent on `omitempty`.

**Review by laplaque — CHANGES_REQUESTED** [review]
## PR Review: feat(streaming): deanonymize DeepSeek `reasoning_content`

### Pipeline status
PASS — all 8 CI checks green (Lint, CodeQL, Test, Security Scan, Benchmark, Build).

### Test coverage verdict
PASS on changed functions per PR body (100/98/100% on `newOpenAIDeanonymizer` / `ProcessDataPayload` / `Flush`), consistent with the six new tests + regression guard.

### Path to Approval (Action Plan)
1. **Restructure reasoning + content as isolated emissions.** Don't mutate the input envelope; construct a fresh `openAIEnvelope` per emission with the delta carrying only the relevant field. Remove the early `return true` after the reasoning branch so a chunk with both fields populated produces two output chunks (reasoning, then content) without cross-contamination. Add a test asserting both fields land in separate output chunks and neither chunk contains the other field's text. See inline comment.

### Verdict
REQUEST_CHANGES

File: internal/anonymizer/streaming_openai.go (hunk 1 above)
Comment (iamclaude697) [reply internal/anonymizer/streaming_openai.go:87]: Addressed in 763163b. Refactored `ProcessDataPayload` so each stream is emitted through a freshly constructed `openAIEnvelope` whose delta carries only the relevant field (`emitReasoning` / `emitContent` helpers). Both branches run unconditionally, so a chunk carrying both fields produces two isolated output chunks — separation is now structural, not `omitempty`-dependent.

Added `TestDeepSeekBothFieldsStructuralIsolation` covering (a) both-field input → two output chunks, (b) reasoning chunk has no content text or content PII, (c) content chunk has no reasoning text or reasoning PII.

Ready for re-review.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by laplaque — CHANGES_REQUESTED** [review]
## PR Review: feat(streaming): deanonymize DeepSeek `reasoning_content` (re-review)

### Pipeline status
PASS — all 8 CI checks green on `763163bc` (Lint, CodeQL actions/go/CodeQL, CI Test under `-race`, Security Scan, Benchmark, Build). Delta-coverage gate passed.

### Code change
The structural-isolation refactor matches the prescription from review [#4260277940](https://github.com/laplaque/ai-anonymizing-proxy/pull/113#pullrequestreview-4260277940):
- `emitReasoning` and `emitContent` each construct a fresh `openAIEnvelope` (no input-envelope mutation).
- Both branches in `ProcessDataPayload` run unconditionally; the early `return true` after reasoning is gone.
- Doc comments on both helpers explicitly state the isolation property and the future-field guard.
- `TestDeepSeekBothFieldsStructuralIsolation` parses every emitted `data:` line and asserts (a) no chunk carries both fields, (b) reasoning chunk leaks no content text/PII, (c) content chunk leaks no reasoning text/PII, plus liveness and originals/no-raw-tokens guards.

The code change is correct. The structural-isolation thread can be resolved.

### Outstanding finding

**MEDIUM — post-refactor baseline-vs-result evidence is missing.**

The PR body's "Quality Gates", "No regressions vs baseline", and "Coverage delta" claims (`make check` green, `go test -race ./...` green, explicit list of 8 existing OpenAI streaming tests confirmed passing, `streaming_openai.go`: `newOpenAIDeanonymizer` 100% / `ProcessDataPayload` 98% / `Flush` 100%) all pertain to the original commit `12ae0af4`. The refactor commit `763163bc` materially changed the function shape — `ProcessDataPayload` was split into two new helpers `emitReasoning` and `emitContent` — without re-stating any of those claims against the refactored code. The refactor commit message describes WHAT changed and announces the new test, but does not document:

- Baseline test state before the refactor was applied locally (which of the existing tests were green before you started)
- Post-refactor state of the full suite + the new `TestDeepSeekBothFieldsStructuralIsolation`
- Updated per-function coverage on the now-different function set (`emitReasoning`, `emitContent` are new — no coverage claim attached; `ProcessDataPayload` shape changed — prior 98% number no longer applies)
- Explicit confirmation that the 8 existing OpenAI streaming tests still pass after the refactor

CI green is external evidence that the suite passed after the push, but that is not a substitute for your own documented baseline-vs-result claim. The PR body needs to show you verified the refactor against baseline, not just that CI ran.

### Path to Approval (Action Plan)
1. Add a PR comment (or amend the refactor commit body) with: (a) baseline test status pre-refactor (the existing 8 OpenAI streaming tests), (b) post-refactor status of those 8 plus the new `TestDeepSeekBothFieldsStructuralIsolation`, all green under `go test -race ./...`, (c) updated per-function coverage on `emitReasoning`, `emitContent`, `ProcessDataPayload`, `Flush` from a local `go tool cover -func=coverage.out` run.

### Verdict
REQUEST_CHANGES

**Comment by iamclaude697** [comment]
## Refactor verification evidence (commit `763163b` vs baseline `12ae0af`)

### (a) Pre-refactor baseline at `12ae0af` (before structural-isolation refactor)

Restored `streaming_openai.go` + `streaming_openai_test.go` from `12ae0af` and ran the 8 existing OpenAI streaming tests under `go test -race`:

```
--- PASS: TestOpenAIStreamingTokenSplit (0.00s)
--- PASS: TestOpenAIStreamingEOFFlush (0.00s)
--- PASS: TestOpenAIStreamingDoneSentinel (0.00s)
--- PASS: TestOpenAIStreamingRoleOnlyChunk (0.00s)
--- PASS: TestOpenAIStreamingFinishReason (0.00s)
--- PASS: TestOpenAIStreamingEmptyChoicesFlush (0.00s)
--- PASS: TestOpenAIStreamingVerboseLogging (0.00s)
--- PASS: TestOpenAIStreamingViaAggregatorDomain (0.01s)
    --- PASS: TestOpenAIStreamingViaAggregatorDomain/openrouter.ai
    --- PASS: TestOpenAIStreamingViaAggregatorDomain/api.portkey.ai
    --- PASS: TestOpenAIStreamingViaAggregatorDomain/api.groq.com
    --- PASS: TestOpenAIStreamingViaAggregatorDomain/api.deepseek.com
    --- PASS: TestOpenAIStreamingViaAggregatorDomain/api.fireworks.ai
    --- PASS: TestOpenAIStreamingViaAggregatorDomain/api.x.ai
    --- PASS: TestOpenAIStreamingViaAggregatorDomain/api.endpoints.anyscale.com
PASS
ok  	ai-anonymizing-proxy/internal/anonymizer	1.115s
```

All 8 existing OpenAI streaming tests + 14 sub-cases green pre-refactor. Working tree restored to HEAD (`763163b`) verified via `git status --short` (clean).

### (b) Post-refactor state at HEAD `763163b` under `go test -race`

Same 8 existing tests **plus** the 6 new DeepSeek tests **plus** the regression guard, all green:

```
--- PASS: TestOpenAIStreamingTokenSplit
--- PASS: TestOpenAIStreamingEOFFlush
--- PASS: TestOpenAIStreamingDoneSentinel
--- PASS: TestOpenAIStreamingRoleOnlyChunk
--- PASS: TestOpenAIStreamingFinishReason
--- PASS: TestOpenAIStreamingEmptyChoicesFlush
--- PASS: TestOpenAIStreamingVerboseLogging
--- PASS: TestDeepSeekReasoningContentDeanonymized
--- PASS: TestDeepSeekReasoningTokenSplit
--- PASS: TestDeepSeekReasoningThenContent
--- PASS: TestDeepSeekReasoningEOFFlush
--- PASS: TestDeepSeekKeepAliveComment
--- PASS: TestDeepSeekBothFieldsStructuralIsolation
--- PASS: TestOpenAIUnchangedByReasoningField
--- PASS: TestOpenAIStreamingViaAggregatorDomain (all 7 sub-cases)
PASS
ok  	ai-anonymizing-proxy/internal/anonymizer	1.114s
```

Full repo under `go test -race ./...`:

```
ok  	ai-anonymizing-proxy/cmd/proxy
ok  	ai-anonymizing-proxy/internal/anonymizer
ok  	ai-anonymizing-proxy/internal/anonymizer/packs
ok  	ai-anonymizing-proxy/internal/config
ok  	ai-anonymizing-proxy/internal/logger
ok  	ai-anonymizing-proxy/internal/management
ok  	ai-anonymizing-proxy/internal/metrics
ok  	ai-anonymizing-proxy/internal/mitm
ok  	ai-anonymizing-proxy/internal/proxy
```

### (c) Updated per-function coverage at `763163b`

`go tool cover -func=coverage.out | grep streaming_openai.go`:

| Function | Coverage |
|----------|---------:|
| `newOpenAIDeanonymizer` | 100.0% |
| `ProcessDataPayload`    | 100.0% |
| `emitReasoning`         |  93.3% |
| `emitContent`           | 100.0% |
| `Flush`                 | 100.0% |

`ProcessDataPayload` moved from 98.0% (pre-refactor) to 100.0% — the prior uncovered branch (the early-return after reasoning) was deleted by the refactor. The two new helpers carry their own coverage; the single uncovered line on `emitReasoning` is the `if flushUpTo == 0 { return }` short-circuit when accumulated reasoning text has not yet crossed the suffix guard, which is exercised by `TestDeepSeekReasoningTokenSplit` and `TestDeepSeekReasoningEOFFlush` for the inverse path but not via this specific early-return shape. Package overall remains at 95.4% (well above the 85% repo gate and 90% delta gate on changed files).

Ready for re-review.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/anonymizer/streaming_openai.go [hunk 2]
```diff
@@ -68,22 +75,70 @@ func (o *openAIDeanonymizer) ProcessDataPayload(payload []byte) bool {
...
+	o.reasoningAccum.WriteString(fragment)
+	accumulated := o.reasoningAccum.String()
+
+	flushUpTo := safeCutPoint(accumulated)
+	if flushUpTo == 0 {
+		return
+	}
+
+	toReplace := accumulated[:flushUpTo]
+	replaced := o.opts.replacer.Replace(toReplace)
```
Comment (laplaque) [inline internal/anonymizer/streaming_openai.go:110]: **HIGH — uncovered guard branch on a PII-handling path.**

Per the [coverage table CC published](https://github.com/laplaque/ai-anonymizing-proxy/pull/113#issuecomment-4417086882), this `if flushUpTo == 0 { return }` early-return is the single uncovered statement that puts `emitReasoning` at 93.3%. CC's analysis: the inverse path (`flushUpTo > 0`) is exercised by `TestDeepSeekReasoningTokenSplit` and `TestDeepSeekReasoningEOFFlush`, but the early-return shape itself is not.

SKILL.md `Path-classification before severity` applies: this helper is on the PII rehydration path, so a missing test on any of its branches is **HIGH** minimum.

**Fix.** Add a test that sends a single reasoning chunk so short that `safeCutPoint(accumulated) == 0` — typically a fragment shorter than `tokenSuffixLen` or containing only the start of what could be a PII token. Assert (a) no `data:` chunk is written by the early-return, (b) the fragment is held in `reasoningAccum` (subsequent `Flush()` produces the deanonymized output). One-line input change from `TestDeepSeekReasoningEOFFlush`'s pattern; closes 93.3% → 100% and directly exercises the early-return.

**Review by laplaque — CHANGES_REQUESTED** [review]
## PR Review: feat(streaming): deanonymize DeepSeek `reasoning_content` (third pass)

### Prior findings status
- **Review #4260277940 (structural isolation):** addressed by commit `763163bc`. Thread resolved.
- **Review #4260384221 (missing post-refactor baseline-vs-result evidence):** addressed by [evidence comment 4417086882](https://github.com/laplaque/ai-anonymizing-proxy/pull/113#issuecomment-4417086882). Baseline at `12ae0af` (8 OpenAI streaming tests + 14 aggregator sub-cases green), post-refactor at `763163b` (same 8 + 6 new DeepSeek + regression guard + full repo `go test -race ./...` green across 9 packages), per-function coverage table all documented. Comprehensive and clean.

### Pipeline status
PASS — 8/8 CI checks green on `763163bc`.

### New finding from the documented coverage table

**HIGH — uncovered guard branch on a PII-handling path.**

CC's coverage disclosure shows `emitReasoning` at 93.3%. CC names the uncovered statement explicitly: the `if flushUpTo == 0 { return }` short-circuit when accumulated reasoning text has not yet crossed `safeCutPoint`'s suffix-guard threshold. CC's mitigation argument is that the inverse path (`flushUpTo > 0`) is tested by `TestDeepSeekReasoningTokenSplit` and `TestDeepSeekReasoningEOFFlush` — true, but the early-return branch itself is not directly exercised.

Per SKILL.md path-classification: a missing test on an auth/security/validation/PII path is HIGH minimum, regardless of branch shape. The reasoning_content deanonymization helper is squarely on the PII rehydration path; any guard branch on that helper that is not directly tested is a HIGH coverage finding.

### Path to Approval

1. Add a test (e.g., `TestDeepSeekReasoningShortFragmentHeld`) that sends a single reasoning chunk so short that `safeCutPoint(accumulated) == 0` — typically a fragment shorter than `tokenSuffixLen` or containing only the start of what could be a PII token. Assert that no `data:` chunk was emitted (the fragment is held in the accumulator) and that a subsequent Flush emits the full text with any tokens replaced. This closes `emitReasoning` 93.3% → 100% and directly exercises the early-return shape.

### Verdict
REQUEST_CHANGES

File: internal/anonymizer/streaming_openai.go (hunk 1 above)
Comment (iamclaude697) [reply internal/anonymizer/streaming_openai.go:87]: Addressed in a9eac54.

`TestDeepSeekReasoningShortFragmentHeld` directly exercises the `if flushUpTo == 0 { return }` early-return: a single reasoning chunk of length `≤ tokenSuffixLen` (31 of 33 bytes max), asserts no mid-stream `data:` chunk was written and that the fragment is held in `reasoningAccum` until `Flush()` emits the synthetic chunk with the replaced token.

While verifying coverage I also found the `verbose && toReplace != replaced` log line in `emitReasoning` was the actual statement keeping the function at 93.3% (the early-return statement was already covered by `TestDeepSeekReasoningEOFFlush`'s flow when the cover tool counted the block, but the verbose-log line was the residual gap). Added `TestDeepSeekReasoningVerboseLogging` — mirror of the existing content-path test — to close it.

Final per-function coverage on `streaming_openai.go`:

| Function | Coverage |
|----------|---------:|
| `newOpenAIDeanonymizer` | 100.0% |
| `ProcessDataPayload`    | 100.0% |
| `emitReasoning`         | **100.0%** |
| `emitContent`           | 100.0% |
| `Flush`                 | 100.0% |

Local `make check` exit 0, `go test -race ./...` green across all 9 packages.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by laplaque — APPROVED** [review]
## PR Review: feat(streaming): deanonymize DeepSeek `reasoning_content` — APPROVED

### All prior findings addressed

- **Review #4260277940 (structural isolation):** addressed by `763163bc`. Thread resolved.
- **Review #4260384221 (missing post-refactor baseline-vs-result evidence):** addressed by [evidence comment 4417086882](https://github.com/laplaque/ai-anonymizing-proxy/pull/113#issuecomment-4417086882) — comprehensive baseline at `12ae0af`, post-refactor at `763163b`, per-function coverage table.
- **Review #4260421792 (uncovered guard branch on PII path):** addressed by `a9eac548` with two new tests:
  - `TestDeepSeekReasoningShortFragmentHeld` — directly exercises `if flushUpTo == 0 { return }` with a fragment whose `len ≤ tokenSuffixLen`; asserts no mid-stream emit and exactly one Flush emission (the count==1 check is a nice precision touch).
  - `TestDeepSeekReasoningVerboseLogging` — bonus mirror of the content-side verbose test for the reasoning path.

### Pipeline status
PASS — 8/8 CI checks green on `a9eac548` (Lint, CodeQL Analyze actions/go/CodeQL, Test under `-race -coverprofile`, Security Scan, Benchmark, Build).

### Coverage
The commit message documents `streaming_openai.go` at 100% across all five functions (`newOpenAIDeanonymizer`, `ProcessDataPayload`, `emitReasoning`, `emitContent`, `Flush`). Combined with the prior evidence comment, the baseline (93.3% on `emitReasoning`) → post-change (100%) delta is on the record.

### Verdict
APPROVE — ready to merge.
